File: src/core/Options/Handlers/Details.cs
using System;
using System.Threading;
using System.Threading.Tasks;
using core.Account;
using core.Shared;
using core.Shared.Adapters.Brokerage;

namespace core.Options
{
    public class Details
    {
        public class Query : RequestWithUserId<CommandResponse<OwnedOptionView>>
        {
            public Guid Id { get; }

            public Query(Guid id, Guid userId) : base(userId)
            {
                Id = id;
            }
        }

        public class Handler : HandlerWithStorage<Query, CommandResponse<OwnedOptionView>>
        {
            private readonly IAccountStorage _accounts;
            private readonly IBrokerage _brokerage;

            public Handler(
                IAccountStorage accounts,
                IBrokerage brokerage,
                IPortfolioStorage storage
                ) : base(storage)
            {
                _accounts = accounts;
                _brokerage = brokerage;
            }

            public override async Task<CommandResponse<OwnedOptionView>> Handle(Query request, CancellationToken cancellationToken)
            {
                var user = await _accounts.GetUser(request.UserId) ?? throw new InvalidOperationException("User not found");
                var option = await _storage.GetOwnedOption(request.Id, request.UserId);
                var chain = await _brokerage.GetOptions(user.State, option.State.Ticker);
                var detail = chain.Success?.FindMatchingOption(
                    strikePrice: option.State.StrikePrice,
                    expirationDate: option.State.ExpirationDate,
                    optionType: option.State.OptionType
                );
                
                var model = new OwnedOptionView(option.State, optionDetail: detail);

                return CommandResponse<OwnedOptionView>.Success(model);
            }
        }
    }
}

File: src/core.fs/Options/Details.fs
module core.fs.Options.Details

    open System
    open core
    open core.Account
    open core.Options
    open core.Shared
    open core.Shared.Adapters.Brokerage
    open core.Shared.Adapters.Storage
    open core.fs

    type Query(optionId:Guid, userId:Guid) =
        member this.OptionId = optionId
        member this.UserId = userId
        
    type Handler(accounts:IAccountStorage, brokerage:IBrokerage, storage:IPortfolioStorage) =
        
        interface IApplicationService
        
        member this.Handle (query:Query) = task {
            let! user = accounts.GetUser(query.UserId)
            
            match user with
            | null ->
                return ServiceResponse<OwnedOptionView>(ServiceError("User not found"))
            | _ ->
                
                let! option = storage.GetOwnedOption(optionId=query.OptionId, userId=query.UserId)
                let! chain = brokerage.GetOptions(user.State, option.State.Ticker)
                let detail = chain.Success.FindMatchingOption(
                    strikePrice = option.State.StrikePrice,
                    expirationDate = option.State.ExpirationDate,
                    optionType = option.State.OptionType
                )
                
                return ServiceResponse<OwnedOptionView>(
                    OwnedOptionView(option.State, optionDetail = detail)
                );      
        }
            
            